runtime/src/tokio.rs

```diff
@@ -433,6 +483,135 @@
 
 impl CryptoRng for Context {}
 
+pub struct Blob {
+    file: fs::File,
+    metrics: Arc<Metrics>,
+}
+
+impl crate::Storage<Blob> for Context {
+    async fn open(&mut self, partition: &str, name: &str) -> Result<Blob, Error> {
+        // Construct the full path
+        let path = self
+            .executor
+            .cfg
+            .storage_directory
+            .join(partition)
+            .join(name);
+
+        // Create the partition directory if it does not exist
+        fs::create_dir_all(path.parent().unwrap())
+            .await
+            .map_err(|_| Error::PartitionCreationFailed(partition.into()))?;
+
+        // Open the file in read-write mode, create if it does not exist
+        let file = fs::OpenOptions::new()
+            .read(true)
+            .write(true)
+            .create(true)
+            .truncate(false)
+            .open(&path)
+            .await
+            .map_err(|_| Error::BlobOpenFailed(partition.into(), name.into()))?;
+
+        self.executor.metrics.open_blobs.inc();
+        Ok(Blob {
+            file,
+            metrics: self.executor.metrics.clone(),
+        })
+    }
+
+    async fn remove(&mut self, partition: &str, name: Option<&str>) -> Result<(), Error> {
+        let path = self.executor.cfg.storage_directory.join(partition);
+        if let Some(name) = name {
+            let blob_path = path.join(name);
+            fs::remove_file(blob_path)
+                .await
+                .map_err(|_| Error::BlobMissing(partition.into(), name.into()))?;
+        } else {
+            fs::remove_dir_all(path)
+                .await
+                .map_err(|_| Error::PartitionMissing(partition.into()))?;
+        }
+        Ok(())
+    }
+
+    async fn scan(&self, partition: &str) -> Result<Vec<String>, Error> {
+        let path = self.executor.cfg.storage_directory.join(partition);
+        let mut entries = fs::read_dir(path)
+            .await
+            .map_err(|_| Error::PartitionMissing(partition.into()))?;
+        let mut blobs = Vec::new();
+        while let Some(entry) = entries.next_entry().await.map_err(|_| Error::ReadFailed)? {
+            let file_type = entry.file_type().await.map_err(|_| Error::ReadFailed)?;
+            if !file_type.is_file() {
+                continue;
+            }
+            if let Some(name) = entry.file_name().to_str() {
+                blobs.push(name.into());
+            }
+        }
+        Ok(blobs)
```

Comment:
If a partition becomes corrupt (such that we can't determine the file type of some blob), we should error loudly.

I am going to remove the `continue` on if `is_file` because that should never happen.